File: Criacionais/Singleton/Singleton02.js
const Carrinho = (function(){
    let instance;

    function createInstance(){
        let produtos = [];

        function addProduto(produto){
            produtos.push(produto);
        }

        function getProduto(){
            return produtos;
        }

        function clearCarrinho(){
            produtos = [];
        }

        return{
            addProduto,
            getProduto,
            clearCarrinho
        };
    }

    return{
        getInstance: function(){
            if(!instance){
                instance = createInstance();
            }
            return instance;
        }
    }
})();

//cliente------------------

const carrinhoInstance1 = Carrinho.getInstance();

console.log("#Carrinho 01")
carrinhoInstance1.addProduto({id: 1, nome: "produto1", preco: 10});
carrinhoInstance1.addProduto({id: 1, nome: "produto1", preco: 10});
carrinhoInstance1.addProduto({id: 1, nome: "produto1", preco: 10});
carrinhoInstance1.addProduto({id: 1, nome: "produto1", preco: 10});

console.log(carrinhoInstance1.getProduto());


const carrinhoInstance2 = Carrinho.getInstance();

carrinhoInstance2.clearCarrinho();

console.log("#Carrinho 02")
carrinhoInstance2.addProduto({id: 2, nome: "produto2", preco: 20});
carrinhoInstance2.addProduto({id: 2, nome: "produto2", preco: 20});
carrinhoInstance2.addProduto({id: 2, nome: "produto2", preco: 20});
carrinhoInstance2.addProduto({id: 2, nome: "produto2", preco: 20});

console.log(carrinhoInstance2.getProduto());

File: Criacionais/Singleton/Singleton02.py
class Carrinho:
    instance = None

    def __init__(self):
        if Carrinho.instance is None:
            Carrinho.instance = self
            self.items = []
        else:
            raise Exception("Esta Classe é um Singleton!")
    
    @staticmethod
    def get_instance():
        if not Carrinho.instance:
            Carrinho()
        return Carrinho.instance
    
    def add_item(self, item):
        self.items.append(item)
    
    def get_items(self):
        return self.items

class Produto:
    def __init__(self, nome, preco):
        self.nome = nome
        self.preco = preco 
    
# Utilizando Singleton
carrinho1 = Carrinho.get_instance()
carrinho2 = Carrinho.get_instance()

# Adicionar itens no Carrinho 1
produto1 = Produto("Camisa", 50.00)
produto2 = Produto("Jaqueta", 100.00)
produto3 = Produto("Anel", 10.00)

carrinho1.add_item(produto1)
carrinho1.add_item(produto2)
carrinho1.add_item(produto3)
carrinho1.add_item(produto2)
carrinho1.add_item(produto3)

# Exibindo itens do Carrinho 1 e 2
print("Carrinho 01:")
for item in carrinho1.get_items():
    print(item.nome, "-", item.preco)
     
# Cliente
carrinho1 = Carrinho.get_instance()
